### PR TITLE
[expo-updates][cli] Fix CLI VCS detection and add workflow override ability

### PR DESCRIPTION
**Pull request by wschurman**
# Why

When there's no `.git` folder uploaded as part of the build (when `requireCommit` is false or not present in eas.json), even though git is installed, we also need to make sure we're working in a valid repo before deciding to use the `GitClient` as opposed to the `NoVCSClient` for workflow detection.

Part 2 of this is adding the ability to override workflow detection so that in EAS Build we can do workflow detection before running prebuild (which invalidates workflow detection since workflow detection depends on the presence of the native folders and their gitignore status).

Closes ENG-11995.

# How

When deciding which vcs system to use to detect workflow, also check if the root path detection command will work.

# Test Plan

For VCS fix:

1. Create a canary project, configure eas and eas-updates
2. `npm pack` expo-updates
3. add to project, use in project's package.json
4. run eas build

before:
https://staging.expo.dev/accounts/real-org-will/projects/test-normal-build-norequirecommit/builds/f8d5512b-32c6-40b6-8842-4d9ccbd1c8f0

after:
https://staging.expo.dev/accounts/real-org-will/projects/test-normal-build-norequirecommit/builds/c7e38a7b-9b0c-4e80-b5d7-85fc6a8c4430

(note that it's still broken, but for a reason related to the second part which will require multiple fixes to fix)

For workflow override:

1. `yarn link expo-updates` in project above locally

```
npx expo-updates runtimeversion:resolve --platform android --workflow managed
npx expo-updates runtimeversion:resolve --platform android --workflow generic
npx expo-updates fingerprint:generate --platform android --workflow managed
npx expo-updates fingerprint:generate --platform android --workflow generic
```

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-updates/utils/src/vcs.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export interface Client {
 }
 
 export default async function getVCSClientAsync(projectDir: string): Promise<Client> {
-  if (await isGitInstalledAsync()) {
+  if (await isGitInstalledAndConfiguredAsync()) {
```
Comment (brentvatne) [inline packages/expo-updates/utils/src/vcs.ts:13]: this seems like something we should extract to a package that we share between eas-cli and expo-updates cli, and potentially other places where we want to do this same thing (or library authors in the community want to)

not a blocker for this pr but something to think about in the future

Comment (wschurman) [reply packages/expo-updates/utils/src/vcs.ts:13]: Agreed.

**Comment by brentvatne** [comment]
is it possible for us to always 100% accurately determine the workflow from a shallow copy of the project, without **.git**? we might have to use some env var that EAS Build sets or whatever the prebuild step does

**Comment by wschurman** [comment]
> is it possible for us to always 100% accurately determine the workflow from a shallow copy of the project, without **.git**? we might have to use some env var that EAS Build sets or whatever the prebuild step does

Yeah, I think that's what w're going to have to do. Pass in the workflow rather than try to detect it. Then, in eas-cli we use that code to do it. In eas-build, the workflow is passed in via the job from eas-cli.

Going to work on that to see how it does.

**Comment by wschurman** [comment]
Edit: nevermind, we need to be able to detect it since this needs to work in a regular project with just expo-updates installed. So maybe the solution is to provide a mechanism for overriding the workflow via a flag in the same way we provide a mechanism for overriding the fingerprint for use in EAS Build.

**Comment by brentvatne** [comment]
got it, so basically we need to do what we're doing in this PR and then provide some mechanism for overriding this behavior with a pre-calculated value

**Comment by wschurman** [comment]
> got it, so basically we need to do what we're doing in this PR and then provide some mechanism for overriding this behavior with a pre-calculated value

I think so. Though we may also just be able to do the workflow calculation step (runtime version calculation / fingerprint calculation) ahead of prebuild. I have to think about why we didn't do it earlier. I think it was because the `POST_INSTALL_HOOK` step can theoretically alter the fingerprint (and even runtime version technically), so we need to do the calculation after that step and that step by definition occurs after PREBUILD and INSTALL_PODS. So yeah, I think we need to do some override mechanism for EAS Build. Working on it now.

**Review by douglowder — APPROVED** [review]

**Pull request merged**